### DISCUSSION TITLE
useSnapshot not rerendering component

### DISCUSSION THREAD

**Discussion by Ghirigoro** [pmndrs/valtio#584] [Q&A]
I'm having trouble with `useSnapshot` not causing a re-render as I would expect:
```ts
const useProposalVotes = (issue: Issue): ProposalVotes => {
  // `ctx` is a proxied value
  const ctx = useStudentAppContext();
  // I would assume that this would cause the hook to re-run when there's any
  // change in any of the children of `ctx.session.dynamic.users`, but it doesn't. 
  const users = useSnapshot(ctx.session.dynamic.users);

  // This however does work, indicating that `users` is changing and the change is
  // detected.
  watch((get) => {
    get(ctx.session.dynamic.users);
    console.log("Change in users detected");
  });

  const votes = issue.proposals.map((p, i) => {
    let v = 0;
    // I reference `users` here, which should cause the hook to re-run.
    for (const key in users) {
      if (users[key].state.groupVote === p.title) v++;
    }
    return v;
  });
  let winnerIndex = -1;
  if (votes[0] > votes[1]) {
    winnerIndex = 0;
  } else if (votes[1] > votes[0]) {
    winnerIndex = 1;
  }
  return { winnerIndex, votes };
};
```
Any thoughts as to why?

**Comment by dai-shi** [comment]
It looks okay.
Trying to create a minimal reproduction would help.
We support `in` operator, but it's rarely used, so there can be some unknown bugs.

**Reply by Ghirigoro** [reply]
Thanks for the quick response. I'll try to put a minimal reproduction together. If not the `in` operator what else would you use in this situation (originally I was iterating with `Object.values` but that didn't work either)?

**Reply by dai-shi** [reply]
`.forEach` should work too. I'd use `.reduce`.
I think `Object.values` should also be supported, but not 100% sure.

**Reply by Ghirigoro** [reply]
I was using `ref` in a few places to avoid duplicate proxies, but I must have missed something. Proxying everything fixed the problem so the issue is elsewhere. Sorry for the false alarm. Is proxying an already proxied object a source of potential problems (aside from a minor performance issue)?

**Reply by dai-shi** [reply]
Proxying a proxied object should be no problem.
